crates/storage/src/column.rs

```diff
@@ -0,0 +1,189 @@
+        ContractsAssetsMerkleMetadata = 12,
+        /// See [`ContractsStateMerkleData`](storage::ContractsStateMerkleData)
+        ContractsStateMerkleData = 13,
+        /// See [`ContractsStateMerkleMetadata`](storage::ContractsStateMerkleMetadata)
+        ContractsStateMerkleMetadata = 14,
+        /// See [`Messages`](crate::tables::Messages)
+        Messages = 15,
+
+        // TODO: Extract the columns below into a separate enum to not mix
+        //  required columns and non-required columns. It will break `MemoryStore`
+        //  and `MemoryTransactionView` because they rely on linear index incrementation.
+
+        // Below are the tables used for p2p, block production, starting the node.
+
+        /// The column id of metadata about the blockchain
+        Metadata = 16,
+        /// See [`Receipts`](crate::tables::Receipts)
+        Receipts = 17,
+        /// See [`FuelBlockSecondaryKeyBlockHeights`](storage::FuelBlockSecondaryKeyBlockHeights)
+        FuelBlockSecondaryKeyBlockHeights = 18,
+        /// See [`SealedBlockConsensus`](crate::tables::SealedBlockConsensus)
+        FuelBlockConsensus = 19,
+        /// Metadata for the relayer
+        /// See [`RelayerMetadata`](fuel_core_relayer::ports::RelayerMetadata)
+        RelayerMetadata = 20,
+
+        // Below are not required tables. They are used for API and may be removed or moved to another place in the future.
+
+        /// The column of the table that stores `true` if `owner` owns `Coin` with `coin_id`
+        OwnedCoins = 21,
+        /// Transaction id to current status
+        TransactionStatus = 22,
+        /// The column of the table of all `owner`'s transactions
+        TransactionsByOwnerBlockIdx = 23,
+        /// The column of the table that stores `true` if `owner` owns `Message` with `message_id`
+        OwnedMessageIds = 24,
+    }
+}
+
+impl Column {
+    /// The total count of variants in the enum.
+    pub const COUNT: usize = <Self as strum::EnumCount>::COUNT;
+
+    /// Returns the `usize` representation of the `Column`.
+    pub fn as_usize(&self) -> usize {
+        self.as_u32() as usize
+    }
+}
+
+impl StorageColumn for Column {
+    fn name(&self) -> &'static str {
+        match self {
+            Column::ForeignColumn(foreign) => foreign.name,
+            variant => variant.into(),
+        }
+    }
+
+    fn id(&self) -> u32 {
+        self.as_u32()
+    }
+}
+
+/// The foreign column is not related to the required tables.
+/// It can be used to extend the database with additional tables.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct ForeignColumn {
+    id: u32,
+    name: &'static str,
+}
+
+impl ForeignColumn {
+    /// Creates the foreign column ensuring that the id and name
+    /// are not already used by the [`Column`] required tables.
+    pub fn new(id: u32, name: &'static str) -> anyhow::Result<Self> {
+        for column in enum_iterator::all::<Column>() {
```

Comment:
Is this a check we should do at compile time (i.e. const eval) rather than every time a ForeignColumn is instantiated?